securedrop_export/disk/new_service.py

```diff
@@ -0,0 +1,120 @@
+    def __init__(self, cli: CLI):
+        self.cli = cli
+
+    def run(self, arg: str) -> Status:
+        """
+        Run export actions.
+        """
+
+    def scan_all_devices(self) -> Status:
+        """
+        Check all connected devices and return current device
+        status.
+        """
+        try:
+            all_devices = self.cli.get_connected_devices()
+            number_devices = len(all_devices)
+
+            if number_devices == 0:
+                return Status.NO_DEVICE_DETECTED
+            elif number_devices > 1:
+                return Status.MULTI_DEVICE_DETECTED
+            else:
+                return self.scan_single_device(all_devices[0])
+
+        except ExportException as ex:
+            logger.error(ex)
+            return Status.DEVICE_ERROR  # Could not assess devices
+
+    def scan_single_device(self, blkid: str) -> Status:
+        """
+        Given a string representing a single block device, see if it
+        is a suitable export target and return information about its state.
+        """
+        try:
+            target = self.cli.get_partitioned_device(blkid)
+
+            # See if it's a LUKS drive
+            if self.cli.is_luks_volume(target):
+
+                # Returns Volume or throws ExportException
+                self.volume = self.cli.get_luks_volume(target)
+
+                # See if it's unlocked and mounted
+                if self.volume.writable:
+                    logger.debug("LUKS device is already mounted")
+                    return Status.DEVICE_WRITABLE
+                else:
+                    # Prompt for passphrase
+                    return Status.DEVICE_LOCKED
+            else:
+                # Might be VeraCrypt, might be madness
+                logger.info("LUKS drive not found")
+
+                # Currently we don't support anything other than LUKS.
+                # In future, we will support TC/VC volumes as well
+                return Status.INVALID_DEVICE_DETECTED
+
+        except ExportException as ex:
+            logger.error(ex)
+            if ex.sdstatus:
+                return ex.sdstatus
+            else:
+                return Status.DEVICE_ERROR
+
+    def unlock_device(self, passphrase: str, volume: Volume) -> Status:
+        """
+        Given provided passphrase, unlock target volume. Currently,
+        LUKS volumes are supported.
+        """
+        if volume:
+            try:
+                self.volume = self.cli.unlock_luks_volume(volume, passphrase)
+
+                if volume.writable:
+                    return Status.DEVICE_WRITABLE
+                else:
+                    return Status.ERROR_UNLOCK_LUKS
+
+            except ExportException as ex:
+                logger.error(ex)
+                return Status.ERROR_UNLOCK_LUKS
+        else:
+            # Trying to unlock devices before having an active device
+            logger.warning("Tried to unlock_device but no current volume detected.")
```

Comment:
In my mind, if we ever got to more continuous polling of device status, and the user ever yanked out their device, `volume` would be nullable (or have some other attribute that indicated it was prematurely removed) and this would be a warning not an error.  However, you're right we're not there yet and rn it's an error.

Reply:
Happy to make the change.

In my mind there's a distinction between `NO_DEVICE_DETECTED` (would lead to "please insert a device!" and would not be an error--there is still a state to continue to from here in your wizard implementation) vs, for example `INVALID_DEVICE` (raise an error, can't do more with this device, it's gotta go.)  So that's why I didn't originally `raise ExportException(NO_DEVICE_DETECTED)` or log it as an error here.

Reply:
Oh, that's a good point you raise. I'm taking the perspective (which is only _a_ perspective, but is better being explicit) that _logging_ is about the module itself and it's usage, and I don't even consider the GUI or end-user context for that. (It would probably be fair to say consider ourselves, developers, as the users of logs.)

I do consider the GUI context during error _handling_, but didn't even consider doing that to make _logging_ decision. What do you think?